### PR TITLE
Support BSON.raw

### PR DESCRIPTION
**Pull request by yizhiheng**

### CODE REVIEW

**Comment by qlikossbuild** [comment]
Hi, and thanks for contributing!

Note: If you are part of Qlik R&D, you can apply to the `open-source` team in our _internal_ GitHub organization.

Before we can handle your pull request we need you to sign our contributor license agreement.
This is a one-time process and enables you to contribute without hassle to any of the qlik-oss repositories in the future.

And it only takes a few minutes! Here are some useful links to get you started:

* [Sign the CLA](https://github.com/qlik-oss/open-source/blob/master/sign-cla.md#sign-the-cla)
* [Rationale behind the CLA](https://github.com/qlik-oss/open-source/blob/master/cla-rationale.md)
* [Qlik R&D Open Source repository](https://github.com/qlik-oss/open-source#open-source-at-qlik-rd)
* [Official Slack team](http://branch.qlik.com/slack)

Having troubles? Feel free to reach out to us either here or on Slack!

**Note:** Once the CLA is sorted, you should rebase/touch this pull request to trigger an update of this CLA status.

File: mongo/find.go [hunk 1]
```diff
@@ -296,6 +296,10 @@ func generateCursor(result interface{}, paginatedField string, shouldSecondarySo
 	} else {
 		typ = val.Type()
 	}
+	if typ.String() == "bson.Raw" {
```
Comment (EtienneDufresne) [inline mongo/find.go]: I think we should support bson.Raw and *bson.Raw.

File: mongo/find.go [hunk 2]
```diff
@@ -296,6 +296,10 @@ func generateCursor(result interface{}, paginatedField string, shouldSecondarySo
...
+		return generateCursorFromBSONRaw(result, paginatedField, shouldSecondarySortOnID)
```
Comment (EtienneDufresne) [inline mongo/find.go]:
```suggestion
		return generateCursorFromBSONRaw(result.(bson.Raw), paginatedField, shouldSecondarySortOnID)
```

File: mongo/find.go [hunk 3]
```diff
@@ -296,6 +296,10 @@ func generateCursor(result interface{}, paginatedField string, shouldSecondarySo
 	} else {
 		typ = val.Type()
```
Comment (EtienneDufresne) [inline mongo/find.go]: You did not change this but can you get rid of this else by defining and setting `typ` like this? `typ := val.Type() `

File: mongo/find.go [hunk 4]
```diff
@@ -296,6 +296,10 @@ func generateCursor(result interface{}, paginatedField string, shouldSecondarySo
...
 	}
+	if typ.String() == "bson.Raw" {
+		return generateCursorFromBSONRaw(result.(bson.Raw), paginatedField, shouldSecondarySortOnID)
+	}
```
Comment (EtienneDufresne) [inline mongo/find.go:295]: Create a generateCursorFromStruct(...) func that is called here.

**Review by EtienneDufresne — COMMENTED** [review]
Don't forget to add integrations tests for bson.Raw and *bson.Raw.

**Review by EtienneDufresne — COMMENTED** [review]
Ideally, this would be supported for the mgo driver do you mind making the change there as well? Please also add component tests.

**Review by EtienneDufresne — COMMENTED** [review]
Do you mind placing the integration tests for each driver in their own package like we did for the implementation?

File: test/integration/mgo_items_store.go [hunk 5]
```diff
@@ -19,6 +19,7 @@ type (
 	// Store allows operations on items.
 	Store interface {
 		Create(i *Item) (*Item, error)
+		RemoveAll() error
```
Comment (EtienneDufresne) [inline test/integration/mgo_items_store.go:22]: Can you add a FindBSONRaw implementation and add a test?

Comment (yizhiheng) [reply test/integration/mgo_items_store.go:22]: will do!

**Review by EtienneDufresne — APPROVED** [review]

**Pull request merged**